Fix :is(:host) both as a subject and as an ancestor

mdubet opened the pull request:
#### b2e941c46d19000ece893d2f79e9ff52fda755a5
<pre>
Fix :is(:host) both as a subject and as an ancestor
<a href="https://bugs.webkit.org/show_bug.cgi?id=264628">https://bugs.webkit.org/show_bug.cgi?id=264628</a>
<a href="https://rdar.apple.com/118582384">rdar://118582384</a>

Reviewed by Antti Koivisto.

As ancestor: we always follow the Shadow DOM host, and check a posteriori
that we have successfully matched an :host pseudoclass.

As subject: rules containg compound :host will not end up in the host bucket
but in the universal bucket of the shadow host, we mark those as such and collect them when needed.

* LayoutTests/TestExpectations:
* LayoutTests/fast/css/host-matching-expected.html: Added.
* LayoutTests/fast/css/host-matching.html: Added.
* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::RareData::deepCopy const):
(WebCore::CSSSelector::isHostPseudoClass const):
* Source/WebCore/css/CSSSelector.h:
* Source/WebCore/css/SelectorChecker.cpp:
(WebCore::SelectorChecker::matchHostPseudoClass const):
(WebCore::localContextForParent):
(WebCore::SelectorChecker::matchRecursively const):
(WebCore::SelectorChecker::checkOne const):
* Source/WebCore/css/SelectorChecker.h:
* Source/WebCore/css/parser/CSSSelectorParser.cpp:
(WebCore::CSSSelectorParser::prependTypeSelectorIfNeeded):
* Source/WebCore/css/parser/MutableCSSSelector.cpp:
(WebCore::MutableCSSSelector::isHostPseudoSelector const): Deleted.
* Source/WebCore/css/parser/MutableCSSSelector.h:
(WebCore::MutableCSSSelector::isHostPseudoClass const):
* Source/WebCore/style/ElementRuleCollector.cpp:
(WebCore::Style::ElementRuleCollector::matchHostPseudoClassRules):
* Source/WebCore/style/RuleSet.cpp:
(WebCore::Style::hasHostPseudoClassSubjectInSelectorList):
(WebCore::Style::isHostSelectorMatchingInShadowTree):
(WebCore::Style::RuleSet::addRule):
* Source/WebCore/style/RuleSet.h:
(WebCore::Style::RuleSet::hasHostPseudoClassRulesInUniversalBucket const):

Canonical link: <a href="https://commits.webkit.org/281963@main">https://commits.webkit.org/281963@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/88ebf9b464735d8eb76ae11e876a25b100bbac23

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/61537 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/40885 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/14109 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/65478 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/12076 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/63656 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/48563 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/12348 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/49682 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/8384 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/64606 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/37985 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/53296 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/30514 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/34653 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/10528 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/10986 "Build is being retried. Recent messages:Printed configuration; Running apply-patch; Checked out pull request; Compiled WebKit") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/56460 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/10824 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/67208 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/5470 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/10593 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/57064 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/5495 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/53257 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/57277 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13716 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/4515 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/36690 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/37775 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/38869 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/37517 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/css/CSSSelector.cpp:
```diff
@@ -864,4 +864,24 @@ bool CSSSelector::hasExplicitPseudoClassScope() const
+    }
+
+    return false;
+}
```
Are this and the call sites correct for `:not(:host)` and all other pseudo classes taking selector list arguments? Should this be limited to `:is()` and `:where()`?

mdubet replied:
It's limited to `:is() :where() (and webkit-any)` now

anttijk commented on Source/WebCore/css/SelectorChecker.cpp:
```diff
@@ -702,9 +702,8 @@ bool SelectorChecker::checkOne(CheckingContext& checkingContext, const LocalCont
         bool isPseudoElement = selector.match() == CSSSelector::Match::PseudoElement;
-        if (!isHostPseudoClass && !isPseudoElement)
+        if (!selector.isOrContainsHostPseudoClass() && !isPseudoElement)
             return false;
```
I'm not sure this would be correct with `:is(:host(foo))` etc. It is not obvious to me why can't do normal recursive matching in this case too and have to dig up the `:host` from inside `:is()`. This is not how SelectorChecker generally works.

nt1m commented on Source/WebCore/css/CSSSelector.cpp:
```diff
@@ -866,4 +866,28 @@ bool CSSSelector::hasExplicitPseudoClassScope() const
+        return true;
+
+    if (match() == Match::PseudoClass) {
+        if (pseudoClass() == PseudoClass::Is || pseudoClass() == PseudoClass::Where || pseudoClass() == PseudoClass::WebKitAny) {
```
I wouldn't change the behavior of `WebKitAny`

nt1m commented on Source/WebCore/css/CSSSelector.cpp:
```diff
@@ -866,4 +866,28 @@ bool CSSSelector::hasExplicitPseudoClassScope() const
+    return match() == Match::PseudoClass && pseudoClass() == PseudoClass::Host;
+}
+
+bool CSSSelector::isOrSimilarToHostPseudoClass() const
```
Would `matchesHostPseudoClass()`  or `containsMatchForHostPseudoClass()` be more descriptive?

mdubet replied:
Function removed

anttijk commented on Source/WebCore/css/SelectorChecker.cpp:
```diff
@@ -269,7 +269,7 @@ static SelectorChecker::LocalContext localContextForParent(const SelectorChecker
 
     // Move to the shadow host if matching :host and the parent is the shadow root.
-    if (context.selector->match() == CSSSelector::Match::PseudoClass && context.selector->pseudoClass() == CSSSelector::PseudoClass::Host) {
+    if (context.selector->isOrSimilarToHostPseudoClass()) {
```
I still don't think this is right conceptually as `:is()` may contain other things besides `:host` we don't know if we got here by matching a `:host` or not. We should move to the host scope when we try to match`:host` inside.

Maybe in

```
        case CSSSelector::PseudoClass::Is:
        case CSSSelector::PseudoClass::Where:
        case CSSSelector::PseudoClass::WebKitAny:
```

branch we could move to the host element similar to localContextForParent() if we see `PseudoElement::Host` as the rightmost component of the subselector.

mdubet replied:
Good point. I've changed the fix.

anttijk commented on Source/WebCore/css/CSSSelector.h:
```diff
@@ -157,6 +157,7 @@ class CSSSelector {
     bool matchesPseudoElement() const;
     bool isSiblingSelector() const;
     bool isAttributeSelector() const;
+    bool isHostPseudoClass() const;
```
We should add helpers like `isPseudoClass(PseudoClass)`,` isPseudoElement(PseudoElement)` etc. at some point

anttijk commented on Source/WebCore/style/ElementRuleCollector.cpp:
```diff
@@ -323,12 +323,21 @@ void ElementRuleCollector::matchHostPseudoClassRules(CascadeLevel level)
+        MatchRequest hostMatchRequest { *shadowRules, ScopeOrdinal::Shadow };
+        collectMatchingRulesForList(&rules, hostMatchRequest);
+
+    };
```
extra empty line

anttijk commented on Source/WebCore/style/RuleSet.h:
```diff
@@ -115,6 +115,7 @@ class RuleSet : public RefCounted<RuleSet> {
     bool hasAttributeRules() const { return !m_attributeLocalNameRules.isEmpty(); }
     bool hasUserAgentPartRules() const { return !m_userAgentPartRules.isEmpty(); }
     bool hasHostPseudoClassRulesMatchingInShadowTree() const { return m_hasHostPseudoClassRulesMatchingInShadowTree; }
+    bool mustCheckShadowHostUniversalBucket() const { return m_mustCheckShadowHostUniversalBucket; }
```
maybe name this the same style as a the rest (hasSomething) instead of what must be done.

anttijk commented on Source/WebCore/style/RuleSet.cpp:
```diff
@@ -75,13 +75,31 @@ static unsigned rulesCountForName(const RuleSet::AtomRuleMap& map, const AtomStr
 
+// FIXME: Maybe we can unify both following functions
+
+static bool maybeHostSubjectInSelectorList(const CSSSelectorList* selectorList)
```
Why "maybe"? This function seems to answer the question definitely ("has").

mdubet committed:
Fix :is(:host) both as a subject and as an ancestor
https://bugs.webkit.org/show_bug.cgi?id=264628
rdar://118582384

Reviewed by Antti Koivisto.

As ancestor: we always follow the Shadow DOM host, and check a posteriori
that we have successfully matched an :host pseudoclass.

As subject: rules containg compound :host will not end up in the host bucket
but in the universal bucket of the shadow host, we mark those as such and collect them when needed.

* LayoutTests/TestExpectations:
* LayoutTests/fast/css/host-matching-expected.html: Added.
* LayoutTests/fast/css/host-matching.html: Added.
* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::RareData::deepCopy const):
(WebCore::CSSSelector::isHostPseudoClass const):
* Source/WebCore/css/CSSSelector.h:
* Source/WebCore/css/SelectorChecker.cpp:
(WebCore::SelectorChecker::matchHostPseudoClass const):
(WebCore::localContextForParent):
(WebCore::SelectorChecker::matchRecursively const):
(WebCore::SelectorChecker::checkOne const):
* Source/WebCore/css/SelectorChecker.h:
* Source/WebCore/css/parser/CSSSelectorParser.cpp:
(WebCore::CSSSelectorParser::prependTypeSelectorIfNeeded):
* Source/WebCore/css/parser/MutableCSSSelector.cpp:
(WebCore::MutableCSSSelector::isHostPseudoSelector const): Deleted.
* Source/WebCore/css/parser/MutableCSSSelector.h:
(WebCore::MutableCSSSelector::isHostPseudoClass const):
* Source/WebCore/style/ElementRuleCollector.cpp:
(WebCore::Style::ElementRuleCollector::matchHostPseudoClassRules):
* Source/WebCore/style/RuleSet.cpp:
(WebCore::Style::hasHostPseudoClassSubjectInSelectorList):
(WebCore::Style::isHostSelectorMatchingInShadowTree):
(WebCore::Style::RuleSet::addRule):
* Source/WebCore/style/RuleSet.h:
(WebCore::Style::RuleSet::hasHostPseudoClassRulesInUniversalBucket const):

Canonical link: https://commits.webkit.org/281963@main
